envision/web/src/components/simulation.js

```diff
@@ -149,19 +149,25 @@ export default function Simulation({
     setScene(scene_);
   };
 
+  const sleep = (milliseconds) => {
+    return new Promise((resolve) => setTimeout(resolve, milliseconds));
+  };
+
   // State subscription
   useEffect(() => {
     let stopPolling = false;
     (async () => {
       const it = client.worldstate(simulationId);
-      let wstate_and_time = await it.next();
-      while (!wstate_and_time.done && playing) {
+      let wstate_and_time;
+      if (playing) wstate_and_time = await it.next();
+      while (!stopPolling && playing && !wstate_and_time.done) {
         let wstate, elapsed_times;
         [wstate, elapsed_times] = wstate_and_time.value;
-        if (!stopPolling) {
-          setWorldState(wstate);
-          onElapsedTimesChanged(...elapsed_times);
-        }
+        setWorldState(wstate);
+        onElapsedTimesChanged(...elapsed_times);
+
+        // play the frames every 50ms
+        await sleep(50);
```

Comment:
Could you please point me to where the rate that Smarts is sending is controlled?  (I saw your comment that it's also 20fps, but I'm just curious to look at it.)

Reply:
yes, SMARTS is sending frame to envision at the following line: https://github.com/huawei-noah/SMARTS/blob/master/smarts/core/smarts.py#L970   `self._envision.send(state)`

Here smarts is just sending the frame whenever an episode is done. 20fps is just an estimate. Since when I use 20fps, the visulation looks normal when transitioning from before pause -> pause -> after resume. If I use a higher fps ( sleep less than 50ms) then i can see the vehicle looks like it is driving way faster after resume for a period of time. Then slows down again when reached the latest frame